iohk/UpdateLogic.hs

```diff
@@ -0,0 +1,344 @@
+  obj <- getObject repo travisRef True
+  case obj of
+    Just (ObjBlob (Blob { blobGetContent = content })) -> do
+      return $ Just content
+    _ -> return Nothing
+
+realFindInstallers :: HasCallStack => T.Text -> (T.Text, T.Text) -> Managed InstallersResults
+realFindInstallers daedalus_rev keys = do
+  daedalus_version <- liftIO $ do
+    contentMaybe <- readFileFromGit daedalus_rev ".travis.yml" "daedalus" "https://github.com/input-output-hk/daedalus"
+    case contentMaybe of
+      Just content -> do
+        case extractVersionFromTravis content of
+          Just version -> return version
+          Nothing -> fail "unable to find daedalus version in .travis.yml"
+      _ -> fail ".travis.yml not found"
+  obj <- liftIO $ fetchGithubStatus "input-output-hk" "daedalus" daedalus_rev
+  -- https://api.travis-ci.org/repos/input-output-hk/daedalus/builds/285552368
+  -- this url returns json, containing the short build#
+  tempdir <- mktempdir "/tmp" "iohk-ops"
+  results <- liftIO $ mapM (findInstaller daedalus_rev daedalus_version (T.pack $ FP.encodeString tempdir) keys) (statuses obj)
+  let
+    findTravis :: [ (Maybe GlobalResults, Maybe CiResult) ] -> CiResult
+    findTravis ((_, Just tr@TravisResult {}) : _) = tr
+    findTravis (_ : rest) = findTravis rest
+    findAppveyor :: [ (Maybe GlobalResults, Maybe CiResult) ] -> CiResult
+    findAppveyor ((_, Just av@AppveyorResult {}) : _) = av
+    findAppveyor (_ : rest) = findAppveyor rest
+    findGlobal :: [ (Maybe GlobalResults, Maybe CiResult) ] -> GlobalResults
+    findGlobal ((Just gr@GlobalResults {}, _) : _) = gr
+    findGlobal (_ : rest) = findGlobal rest
+  _ <- proc "ls" [ "-ltrha", T.pack $ FP.encodeString tempdir ] mempty
+  return $ InstallersResults (findTravis results) (findAppveyor results) (findGlobal results)
+
+fetchOrClone :: TextPath -> RepoUrl -> IO (Git SHA1)
+fetchOrClone localpath url = do
+  res <- isRepo (FP.decodeString $ T.unpack localpath)
+  case res of
+    True -> do
+      fetchRepo localpath url
+    False -> do
+      exitCode <- proc "git" [ "clone", "--mirror", url, localpath ] mempty
+      case exitCode of
+        ExitSuccess   -> fetchRepo localpath url
+        ExitFailure _ -> error "cant clone repo"
+
+fetchRepo :: T.Text -> T.Text -> IO (Git SHA1)
+fetchRepo localpath url = do
+  let
+    fetcher :: Managed (Git SHA1)
+    fetcher = do
+      pushd  $ FP.decodeString $ T.unpack localpath
+      exitCode <- proc "git" [ "fetch", url ] mempty
+      case exitCode of
+        ExitSuccess -> do
+          liftIO $ putStrLn "fetched"
+          repo <- liftIO $ openRepo $ FP.decodeString $ T.unpack localpath
+          pure repo
+        ExitFailure _ -> error "cant fetch repo"
+  with fetcher $ \res -> pure res
+
+processDarwinBuild :: T.Text -> T.Text -> T.Text -> BuildId -> (T.Text, T.Text) -> T.Text -> IO (GlobalResults, CiResult)
+processDarwinBuild daedalus_rev daedalus_version tempdir buildId (winKey, macosKey) ciUrl = do
```

Comment:
I'd prefer if something safer than `(T.Text, T.Text)` were used for the Windows and macOS strings. Somebody may easily mix up the order while refactoring in the future.

There are more places where this same type is used for that purpose, and they should be addressed too.